pkg/lib/serverTime.js

```diff
@@ -321,13 +330,131 @@
+
+        function ensure_sourcedir() {
+            function add_sourcedir(data) {
+                if (data.indexOf("sourcedir /etc/chrony-cockpit-sources") == -1)
+                    data += "\n# Sources added by the Cockpit web console\nsourcedir /etc/chrony-cockpit-sources\n";
+                return data;
+            }
+            return cockpit.file("/etc/chrony.conf", { superuser: "try" }).modify(add_sourcedir);
+        }
+
+        return cockpit.spawn(["mkdir", "-p", "/etc/chrony-cockpit-sources"], { superuser: "try" })
+                .then(() => {
+                    if (enabled)
+                        enabled_file.replace(text).then(() => disabled_file.replace(null)).then(ensure_sourcedir);
+                    else
+                        disabled_file.replace(text).then(() => enabled_file.replace(null));
+                });
+    }
+
+    self.get_custom_ntp = function () {
+        /* We support editing the configuration of systemd-timesyncd
+         * or chronyd, by dropping a config file snippet that we
+         * assume full control over.
+         */
+
+        return get_timesync_backend().then(backend => {
+            if (backend == "systemd-timesyncd")
+                return get_custom_ntp_timesyncd();
+
+            if (backend == "chronyd")
+                return get_custom_ntp_chronyd();
+
+            return Promise.resolve({ supported: false, servers: [], enabled: false });
+        });
+    };
+
+    self.set_custom_ntp = function (servers, enabled) {
+        return get_timesync_backend().then(backend => {
+            if (backend == "systemd-timesyncd")
+                return set_custom_ntp_timesyncd(servers, enabled);
+
+            if (backend == "chronyd")
+                return set_custom_ntp_chronyd(servers, enabled);
+        });
+    };
+
+    self.check_timesync_backend = function () {
+        function wait_enabled(proxy, name) {
+            return new Promise((resolve, reject) => {
+                function check() {
+                    if (proxy.enabled == true) {
+                        proxy.removeEventListener("changed", check);
+                        window.clearTimeout(timeout);
+                        resolve();
+                    }
+                }
+                const timeout = window.setTimeout(() => {
+                    proxy.removeEventListener("changed", check);
+                    reject(cockpit.format(_("Unexpected OS configuration.  The $0 has not been enabled. Your changes might not have had any effect."), name));
+                }, 10000);
+                proxy.addEventListener("changed", check);
+                check();
+            });
+        }
+
+        return get_timesync_backend().then(backend => {
+            if (backend == "systemd-timesyncd")
+                return wait_enabled(timesyncd_service, "systemd-timesyncd.service");
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17992-20221213-115634-adc42487-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)